Properly check whether a transaction was previously in a block or not

ptschip opened the pull request:
The current method is somewhat undesirable in that it requires a lock on cs_utxo (which by the way is missing to begin with). Furthermore it is not accurate in that you can't count on coins actually being in the cache as proof of previous existence of transaction because coins can be uncached when the mempool or orphanpool is full.

We also get the benefit here of reducing the scope of cs_main when receiving transaction inventory.

ptschip committed:
Track recent transactions in blocks and do not re-request them

*Thanks to Matt Corallo (Sorry, I wasn't able to do a proper cherry
pick here so couldn't give proper attibution.)

It is possible that transaction announcements arrive after a block
is processed and we do not want to re-request those txns. In the past
we used the existence of coins in the coins cache as evidence of these
transactions however in BU those coins will likely no longer be in cache
anyway once the block is handled, due to the way we trim the dbcache,
and furthermore we would have required a lock on the utxo to be
maintained (which we were in fact missing).  This approach relies
on a separate lock on cs_recentRejects which avoids any contention
and also ensures we don't ask for those txns again.

ptschip committed:
Simplify the logic in AlreadyHave and make it more readable

ptschip committed:
Add DbgAssert to recentRejects in AlreadyHave

ptschip committed:
Split AlreadyHave() into AlreadyHaveTx() and AlreadyHaveBlock()

This is easier to maintain and prevents us from requiring cs_main
for AlreadyHaveTx().

ptschip committed:
Reduce scope of cs_main when we receive INV messages

We only have to lock cs_main if we received block inventory, which
is in fact rare now.  Most INV's are for transactions and we can
save a great deal of cs_main locking here.

ptschip committed:
Reduce scope of cs_recentRejects

Each lock in AlreadyHaveTx is now taken outside the scope of any of
the other locks so that we don't have to worry now or in the future
about locking orders.

sickpig commented on src/main.cpp:
```diff
@@ -203,9 +204,23 @@ std::map<uint256, NodeId> mapBlockSource;
  * Memory used: 1.7MB
  */
-boost::scoped_ptr<CRollingBloomFilter> recentRejects;
+boost::scoped_ptr<CRollingBloomFilter> recentRejects GUARDED_BY(cs_recentRejects);
```
why not using `std::unique_ptr` like we did below for `txn_recently_in_block`?

ptschip replied:
true, i'll change that.

sickpig commented on src/main.cpp:
```diff
@@ -4441,10 +4468,14 @@ bool InitBlockIndex(const CChainParams &chainparams)
+    {
+        LOCK(cs_recentRejects);
+        recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
+        txn_recently_in_block.reset(new CRollingBloomFilter(16000 /* just a few block's worth */, 0.000001));
```
couldn't we parameterize 16000? since it is correlated w/ the max block size wonder if we could a function of it rather than a magic number.

ptschip replied:
yes your'e right, and it should be probably sized something like  ( EB * (avg txns in a block))

sickpig commented on src/main.cpp:
```diff
@@ -4947,15 +4978,12 @@ std::string GetWarnings(const std::string &strFor)
-bool AlreadyHave(const CInv &inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+bool AlreadyHaveTx(const CInv &inv)
 {
-    AssertLockHeld(cs_main);
```
this is definitely a very good thing. reducing the amount of `cs_main` locking is very important.

sickpig commented on src/main.cpp:
```diff
@@ -5572,19 +5600,17 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
-            LOG(NET, "got inv: %s  %s peer=%d\n", inv.ToString(), fAlreadyHave ? "have" : "new", pfrom->id);
-
+            const CInv &inv = vInv[nInv];
             if (inv.type == MSG_BLOCK)
```
Shouldn't we add also `... && (inv.type == MSG_XTHINBLOCK) && (inv.type == MSG_THINBLOCK)` to the `if` condition?

ptschip replied:
No I don't so.. the caller will know if it's a block or txn inventory, and it doesn't matter if its a BLOCK/THINBLOCK/XTHIN here, we only use  the inv to get the hash...probably would be clearer now to get rid of passing the inv and just use the hash.

ptschip replied:
Sorry, I completely misread your comment.  Yeah, here there is no possibility of thinblock or xthinblock inv types here...we never actually send or received an INV for a thinblock/xthin.  We only use such a INV for sending the get_xthin message.

ptschip replied:
You bring a good point here...we should really check specifically for a TX inventory type and then if we don't get either the block or tx inv then we should probably give them a DOS () for sending us invalid inv types.  But that's for another PR.  I"ll make a note of it.

ptschip replied:
opened PR #1084 for this issue.

ptschip committed:
swap boost scoped_ptr for std::unique_ptr

ptschip committed:
Use excessive block size to calculate the number of bloom filter elements

For transactions recently in a block we want to automatically update
the number of elements in the bloom filter to be dependant on the size
of EB. Although we could in theory receive a larger block this calculation
should be sufficent and it does not have to be 100% accurate. The worst
case being we end up asking for a few transations that we already have.
But in general the filtering done here should prevent all transactions
from being asked for and downloaded twice.

sickpig commented on src/main.cpp:
```diff
@@ -203,9 +204,23 @@ std::map<uint256, NodeId> mapBlockSource;
-boost::scoped_ptr<CRollingBloomFilter> recentRejects;
+std::unique_ptr<CRollingBloomFilter> recentRejects GUARDED_BY(cs_recentRejects);
 uint256 hashRecentRejectsChainTip;
 
```
cool

sickpig commented on src/main.cpp:
```diff
@@ -4441,10 +4468,20 @@ bool InitBlockIndex(const CChainParams &chainparams)
+        // We hold a maximum of two blocks worth of data in the event that two blocks
+        // are mined very close in time. But in general we only need one block of data.
+        uint32_t nElements = 2 * excessiveBlockSize / 500;
+        txn_recently_in_block.reset(new CRollingBloomFilter(nElements, 0.000001));
```
good addition
